source/client_channel_handler.c

```diff
@@ -887,21 +916,38 @@ uint16_t mqtt_create_request(
 
 void mqtt_request_complete(struct aws_mqtt_client_connection *connection, int error_code, uint16_t packet_id) {
 
-    struct aws_hash_element *elem = NULL;
-
     AWS_LOGF_TRACE(
         AWS_LS_MQTT_CLIENT,
         "id=%p: message id %" PRIu16 " completed with error code %d, removing from outstanding requests list.",
         (void *)connection,
         packet_id,
         error_code);
+
+    bool found_request = false;
+    aws_mqtt_op_complete_fn *on_complete = NULL;
+    void *on_complete_ud = NULL;
+
     { /* BEGIN CRITICAL SECTION */
         mqtt_connection_lock_synced_data(connection);
+        struct aws_hash_element *elem = NULL;
         aws_hash_table_find(&connection->synced_data.outstanding_requests_table, &packet_id, &elem);
+        if (elem != NULL) {
+            found_request = true;
+
+            struct aws_mqtt_request *request = elem->value;
+            on_complete = request->on_complete;
+            on_complete_ud = request->on_complete_ud;
+
+            /* clean up request resources */
+            aws_hash_table_remove_element(&connection->synced_data.outstanding_requests_table, elem);
```

Comment:
previously, we'd release the lock, invoke the completion callback, re-acquire the lock and THEN try to remove the aws_hash_element. If the hash-table was mutated in the meantime we'd be in a bad place.

Now we grab the lock, free everything, release the lock, then invoke the completion callbacks